Development/nmos/control_protocol_methods.cpp

```diff
@@ -0,0 +1,550 @@
+
+            auto descriptors = value::array();
+            nmos::find_members_by_class_id(resources, resource, class_id, include_derived, recurse, descriptors.as_array());
+
+            return make_control_protocol_response(handle, { nc_method_status::ok }, descriptors);
+        }
+
+        // NcClassManager methods implementation
+        // Get a single class descriptor
+        web::json::value get_control_class(nmos::resources&, nmos::resources::iterator, int32_t handle, const web::json::value& arguments, get_control_protocol_class_handler get_control_protocol_class, get_control_protocol_datatype_handler, slog::base_gate& gate)
+        {
+            const auto& class_id = parse_nc_class_id(nmos::fields::nc::class_id(arguments)); // Class id to search for
+            const auto& include_inherited = nmos::fields::nc::include_inherited(arguments); // If set the descriptor would contain all inherited elements
+
+            slog::log<slog::severities::more_info>(gate, SLOG_FLF) << "Get a single class descriptor: " << "class_id: " << nmos::details::make_nc_class_id(class_id).serialize();
+
+            if (class_id.empty())
+            {
+                // empty class_id
+                return make_control_protocol_error_response(handle, { nc_method_status::parameter_error }, U("empty classId to do GetControlClass"));
+            }
+
+            // note, model mutex is already locked by the outter function, so access to control_protocol_resources is OK...
+
+            const auto& control_class = get_control_protocol_class(class_id);
+            if (!control_class.class_id.empty())
+            {
+                auto& description = control_class.description;
+                auto& name = control_class.name;
+                auto& fixed_role = control_class.fixed_role;
+                auto properties = control_class.properties;
+                auto methods = control_class.methods;
+                auto events = control_class.events;
+
+                if (include_inherited)
+                {
+                    auto inherited_class_id = class_id;
+                    inherited_class_id.pop_back();
+
+                    while (!inherited_class_id.empty())
+                    {
+                        const auto& inherited_control_class = get_control_protocol_class(inherited_class_id);
+                        {
+                            for (const auto& property : inherited_control_class.properties.as_array()) { web::json::push_back(properties, property); }
+                            for (const auto& method : inherited_control_class.methods.as_array()) { web::json::push_back(methods, method); }
+                            for (const auto& event : inherited_control_class.events.as_array()) { web::json::push_back(events, event); }
+                        }
+                        inherited_class_id.pop_back();
```

Comment:
This is not clear to me: if `pop_back` is used this assumes that `inherited_class_id` has a specific order. But `get_control_protocol_class()` as implemented in control_protocol_handlers.cpp:11-24  uses a `find`-method which could deliver any position.

Reply:
class_id is The idea here is that we walk the class_id

Reply:
The `inherited_class_id` reflects the ancestry of the class being identified.

> A class id field is either a definition index or an authority key. A definition index is an ordinal that starts at 1 for every inheritance level of the control model class tree for example [ 1, 1, 3, 5].
https://specs.amwa.tv/ms-05-02/branches/v1.0-dev/docs/Framework.html#ncclassid

The idea is that if the `include_inherited` is required, we need to walk the `inherited_class_id` backwards to obtain all the inherited classes. At each pop_back we identify the next inherited `class_id`s, by using it to obtain the inherited `control_class` from the `control_protocol_state`'s control_classes **map** (via `get_control_protocol_class`) to build up the `properties, methods, and events`.